src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.AddDebuggerDisplay
+{
+    internal abstract class AbstractAddDebuggerDisplayCodeRefactoringProvider<TTypeDeclarationSyntax, TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_trackingAnnotation = new SyntaxAnnotation();
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var type = await context.TryGetRelevantNodeAsync<TTypeDeclarationSyntax>().ConfigureAwait(false);
+
+            if (type is { })
+            {
+                if (!DeclaresToStringOverride(type)) return;
+            }
+            else
+            {
+                var method = await context.TryGetRelevantNodeAsync<TMethodDeclarationSyntax>().ConfigureAwait(false);
+                if (!IsToStringOverride(method)) return;
+
+                type = method.FirstAncestorOrSelf<TTypeDeclarationSyntax>();
+            }
+
+            if (HasDebuggerDisplayAttribute(type))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new CodeAction(
+                FeaturesResources.Add_DebuggerDisplay,
+                cancellationToken => ApplyAsync(context.Document, type, cancellationToken)));
+        }
+
+        private async Task<Document> ApplyAsync(Document document, TTypeDeclarationSyntax type, CancellationToken cancellationToken)
+        {
+            var syntaxRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var newAttribute = generator
+                .Attribute("System.Diagnostics.DebuggerDisplayAttribute", generator.LiteralExpression("{ToString(),nq}"))
+                .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            syntaxRoot = syntaxRoot.ReplaceNode(type, generator.AddAttributes(type, newAttribute).WithAdditionalAnnotations(s_trackingAnnotation));
```

Comment:
Does the formatter/document change processor kill the `s_trackingAnnotation`, or do I have to remove it manually to avoid the possibility of ending up with two in the same syntax tree after invoking the refactoring twice?

Reply:
you don't have to worry about that.  To explain why:

After you change the tree, many more 'transformation passes' will be done.  These will include, formatting, simplification, etc.  However, after all this is done, this will finally be applied to the workspace as an actual change to an actual file.  After that, the file will be incrementally parsed.  However, it will only ever use the tree used by a previous parse, not any of the trees you produced.  So these trees will never have previous annotations :)